analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -73,21 +73,34 @@ protected override ProgramState PreProcessSimple(SymbolicContext context)
     {
         if (NullDereferenceCandidate(context.Operation.Instance) is { } candidate
             && candidate.Kind == OperationKindEx.ParameterReference
-            && candidate.ToParameterReference() is var parameterReference
-            && !parameterReference.Parameter.Type.IsValueType
-            && !HasObjectConstraint(parameterReference.Parameter)
-            && !parameterReference.Parameter.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_FromServicesAttribute))
+            && candidate.ToParameterReference() is var reference
+            && !reference.Parameter.Type.IsValueType
+            && MissesObjectConstraint(context.State[reference.Parameter])
+            && !reference.Parameter.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_FromServicesAttribute))
+            {
+                var message = context.Operation.Instance.Syntax.FirstAncestorOrSelf<ConstructorInitializerSyntax>() is not null
```

Comment:
That will show me if the parameter is used in the constructor or not, but I actually need it to be more specific: to see if it's inside the constructor initializer.

```cs
public class Constructor : Base
{
    public Constructor(string s)
    {
        _ = s.Length;       // Noncompliant {{Refactor this method to add validation of parameter 's' before using it.}}
    }

    public Constructor(object o) :
        this(o.ToString())  // Noncompliant {{Refactor this constructor to avoid using members of parameter 'o' because it could be null.}}
    {
    }
}
```

So simply knowing that it's in the constructor isn't enough.
Any idea how to do that using only the semantic information? I haven't found anything in the `IParameterSymbol` which shows where it's used.

Reply:
Ah. I missed that part. You can still look in the parent operations, search for an InvocationOperation and check if the TargetMethod is of Kind Constructor. I think this is 100% correct for C# (but maybe I miss something) because `new C(o.ToString())` is an ObjectCreationOperation. The only invocation operations where the target is a constructor should be `base()` and `this()` calls.
I'm not sure about VB though.